OpenRA.Mods.Common/Widgets/TextNotificationsDisplayWidget.cs

```diff
@@ -0,0 +1,157 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2021 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using OpenRA.Mods.Common.Widgets.Logic;
+using OpenRA.Primitives;
+using OpenRA.Widgets;
+
+namespace OpenRA.Mods.Common.Widgets
+{
+	public class TextNotificationsDisplayWidget : Widget
+	{
+		public readonly int RemoveTime = 0;
+		public readonly int ItemSpacing = 4;
+		public readonly int BottomSpacing = 0;
+		public readonly int LogLength = 8;
+		public readonly bool HideOverflow = true;
+
+		public string ChatTemplate = "CHAT_LINE_TEMPLATE";
+		public string SystemTemplate = "SYSTEM_LINE_TEMPLATE";
+		public string MissionTemplate = "SYSTEM_LINE_TEMPLATE";
+		public string FeedbackTemplate = "SYSTEM_LINE_TEMPLATE";
+		readonly Dictionary<TextNotificationPool, Widget> templates = new Dictionary<TextNotificationPool, Widget>();
+
+		readonly List<int> expirations = new List<int>();
+
+		public override Rectangle EventBounds => Rectangle.Empty;
+
+		public override void Initialize(WidgetArgs args)
+		{
+			base.Initialize(args);
+
+			templates.Add(TextNotificationPool.Chat, Ui.LoadWidget(ChatTemplate, null, new WidgetArgs()));
+			templates.Add(TextNotificationPool.System, Ui.LoadWidget(SystemTemplate, null, new WidgetArgs()));
+			templates.Add(TextNotificationPool.Mission, Ui.LoadWidget(MissionTemplate, null, new WidgetArgs()));
+			templates.Add(TextNotificationPool.Feedback, Ui.LoadWidget(FeedbackTemplate, null, new WidgetArgs()));
+		}
+
+		public override void DrawOuter()
+		{
+			if (!IsVisible() || Children.Count == 0)
+				return;
+
+			var lastChildOverflows = Bounds.Height < Children[Children.Count - 1].Bounds.Height;
+
+			if (HideOverflow && lastChildOverflows)
+				Game.Renderer.EnableScissor(GetDrawBounds());
```

Comment:
`GetDrawBounds` looks relately expensive. Would it make sense to wrap this in a `CachedTransform`?

Reply:
Or, hopefully better, can we avoid the scissor completely by only calling a subset of `child.DrawOuter` below?

Reply:
> Or, hopefully better, can we avoid the scissor completely by only calling a subset of `child.DrawOuter` below?

That's a good idea - will look into it. The scissor would still be necessary when there's a single message that overflows because then I want to display at least the part of it that fits in the widget.

Reply:
I thought we were already adjusting the layout to only ever draw full lines?

Reply:
This is only with regards to the extreme case where the latest message overflows our desired widget max height:
![Screenshot_20210927_161707](https://user-images.githubusercontent.com/1355810/134916927-a19e13d5-2e24-48dd-b85a-3c74e3c97a2e.png)
_Scissors in action_

![Screenshot_20210927_161755](https://user-images.githubusercontent.com/1355810/134917001-66c55fe9-2fef-4134-be40-0ef73734eff2.png)
_The full message is visible in the log_

I don't want to be displaying this (or even longer messages if anyone decides to abuse it):
![image](https://user-images.githubusercontent.com/1355810/134917849-84730175-8c36-4618-a5bd-bbec20b7d89a.png)

I am already avoiding the scissor most of the time - it is only ever used if `HideOverflow` is set to `true` and the last child overflows the desired bounds.

Reply:
OK, the scissor bounds are now calculated only once at widget initialization. This is assuming that all templates use the same font which I dare say would be the case 99% of the time.